### PR TITLE
🎉 New Source: PersistIq

### PR DESCRIPTION
**Pull request by Zaimwa9**
## What
This PR adds a new source: [PersistIq api](https://apidocs.persistiq.com/#introduction)

## Comments
Not sure about badge.md

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> New Connector </strong></summary>
<p>

#### Community member or Airbyter
   
- [X] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [X] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [X] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
image
- [ ] Code reviews completed
- [X] Documentation updated 
    - [X] Connector's `README.md`
    - [X] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [X] `docs/SUMMARY.md`
    - [X] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [X] `docs/integrations/README.md`
    - [X] `airbyte-integrations/builds.md`
- [X] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
</p>
</details>

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</p>
</details>

<details><summary> <strong> Connector Generator </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

image

---
This change is [Reviewable](https://reviewable.io/reviews/airbytehq/airbyte/9515)

### CODE REVIEW

**Comment by alafanechere** [comment]
Hi @Zaimwa9, thank you for your contribution! 
Could you please share evidence that the acceptance tests are passing. Could you also please run `./gradlew format` if you did not already?

**Review by alafanechere — CHANGES_REQUESTED** [review]
@Zaimwa9 I tried to run the acceptance tests but they are failing, please make sure they run successfully and I'll go for a second review.

**Comment by Zaimwa9** [comment]
Hey @alafanechere, Indeed, I was looking at the unit tests. I'm working on making them pass but I'm stuck on the read one.
https://airbytehq.slack.com/archives/C01MFR03D5W/p1642324171491000

Would love a bit of help to understand how to pass it :)
Thanks !

**Comment by alafanechere** [comment]
@Zaimwa9 I made some required changes on my [side branch](https://github.com/airbytehq/airbyte/pull/9520) to make the acceptance pass. Could you please grant us permission to push on your fork (guide [here](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests)) so that I can push my commits on it?

File: airbyte-integrations/connectors/source-persistiq/integration_tests/abnormal_state.json [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-persistiq/integration_tests/abnormal_state.json:1]: Please remove this file if your not using the state in your code.

File: airbyte-integrations/connectors/source-persistiq/source_persistiq/bootstrap.md [hunk 2]
```diff
@@ -0,0 +1,18 @@
+# PersistIq
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-persistiq/source_persistiq/bootstrap.md:1]: This file is a level too deep, should rather be in `airbyte-integrations/connectors/source-persistiq/`

File: airbyte-integrations/connectors/source-persistiq/integration_tests/sample_state.json [hunk 3]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-persistiq/integration_tests/sample_state.json:1]: Please remove this file if you're not using the state.

File: airbyte-integrations/connectors/source-persistiq/integration_tests/expected_campaigns_stream.txt [hunk 4]
```diff
@@ -0,0 +1 @@
+{"stream": "campaigns", "data": {"id": "c_N2G2Mx","name": "Streamer Outreach","creator": {"id": "u_3an2Jp","name": "Gabriel Rossmann","email": "gabriel@punctual.cc"},"stats": {"prospects_contacted": 1,"prospects_reached": 1,"prospects_opened": 1,"prospects_replied": 0,"prospects_bounced": 0,"prospects_optedout": 0,"total_prospects": 1}}, "emitted_at": 1629119628000}
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-persistiq/integration_tests/expected_campaigns_stream.txt:1]: This expected data is specific to your PersistIq account, I added the records we have on the Airbyte sandbox account in a commit on my side branch that I'll push to yours once you give us maintainer permissions on your fork.

File: airbyte-integrations/connectors/source-persistiq/source_persistiq/schemas/campaigns.json [hunk 5]
```diff
@@ -0,0 +1,94 @@
+{
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "type": "object",
+  "properties": {
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-persistiq/source_persistiq/schemas/campaigns.json:4]: Please make sure you only use `null` for nullable type, matching the source API spec. e.g: `id` field is probably not nullable.

File: airbyte-integrations/connectors/source-persistiq/spec.json [hunk 6]
```diff
@@ -0,0 +1,17 @@
+{
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-persistiq/spec.json:1]: Duplicate spec.json file, please keep the other one.

File: docs/integrations/sources/persistiq.md [hunk 7]
```diff
@@ -0,0 +1,38 @@
+# Chartmogul
```
Comment (alafanechere) [inline docs/integrations/sources/persistiq.md:1]:
```suggestion
# PersistIq
```
🤔

**Review by alafanechere — CHANGES_REQUESTED** [review]
I made a first quick review of all the boilerplate files and detected some required changes.
Please check my comments.
I'll go for a python code review soon.

File: airbyte-integrations/connectors/source-persistiq/integration_tests/expected_campaigns_stream.txt (hunk 4 above)
Comment (Zaimwa9) [reply airbyte-integrations/connectors/source-persistiq/integration_tests/expected_campaigns_stream.txt:1]: Should be good 
image

File: docs/integrations/sources/persistiq.md (hunk 7 above)
Comment (Zaimwa9) [reply docs/integrations/sources/persistiq.md:1]: Busted, I took the one from the Chartmogul source but didn't to a good job at copying 😅

File: airbyte-integrations/connectors/source-persistiq/source_persistiq/source.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+
+# Basic full refresh stream
+
+
+class PersistiqStream(HttpStream, ABC):
+    def __init__(self, api_key: str, **kwargs):
+        super().__init__(**kwargs)
+        self.api_key = api_key
+
+    url_base = "https://api.persistiq.com/v1/"
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-persistiq/source_persistiq/source.py:22]: Class attribute are usually defined before the init method.

File: airbyte-integrations/connectors/source-persistiq/source_persistiq/source.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+
+
+class Leads(PersistiqStream):
+    primary_key = "id"
+
+    def path(
+        self, **kwargs
+    ) -> str:
+        return "leads"
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-persistiq/source_persistiq/source.py:63]: Please add a space and please run `./gradlew format` to format your code.

File: airbyte-integrations/connectors/source-persistiq/source_persistiq/source.py [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+            return True, None
+        except requests.exceptions.RequestException as e:
+            return False, e
+
+    def streams(self, config: Mapping[str, Any]) -> List[Stream]:
+        auth = NoAuth()
+        return [
+            Users(authenticator=auth, api_key=config["api_key"]),
+            Leads(authenticator=auth, api_key=config["api_key"]),
+            Campaigns(authenticator=auth, api_key=config["api_key"]),
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-persistiq/source_persistiq/source.py:98]:
```suggestion
        return [
            Users(api_key=config["api_key"]),
            Leads(api_key=config["api_key"]),
            Campaigns(api_key=config["api_key"]),
```
`authenticator` already defaults to `NoAuth`

File: airbyte-integrations/connectors/source-persistiq/source_persistiq/spec.json [hunk 11]
```diff
@@ -0,0 +1,19 @@
+{
+  "documentationUrl": "https://docsurl.com",
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-persistiq/source_persistiq/spec.json:2]:
```suggestion
  "documentationUrl": "https://docs.airbyte.io/integrations/sources/persistiq",
```

File: airbyte-integrations/connectors/source-persistiq/unit_tests/test_streams.py [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+def test_request_headers(patch_base_class):
+    stream = PersistiqStream(api_key="mybeautifulkey")
+    expected_headers = {"x-api-key": "mybeautifulkey"}
+    assert stream.request_headers() == expected_headers
+
+
+def test_http_method(patch_base_class):
+    stream = PersistiqStream(api_key="mybeautifulkey")
+    expected_method = "GET"
+    assert stream.http_method == expected_method
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-persistiq/unit_tests/test_streams.py:72]: I don't think your overriding the default method, you can't remove this test if I'm right.

**Review by alafanechere — CHANGES_REQUESTED** [review]
Minor comment and suggestions. Let me know what you think.

**Comment by alafanechere** [comment]
Could you let me know when you give me maintainer permission to push my commits on your branch?

**Comment by alafanechere** [comment]
Thank you @Zaimwa9 for the changes. I'm running the test on our CI now and will let you know how it goes.

**Comment by alafanechere** [comment]
@Zaimwa9 thanks for your contrib, I made minor changes to make the test pass.

**Review by alafanechere — APPROVED** [review]

**Comment by Zaimwa9** [comment]
Thanks a lot, couldn't look at the evolution earlier.

🎉

**Pull request merged**